src/client/opamTreeCommand.ml

```diff
@@ -378,19 +378,35 @@ let print_solution st new_st missing solution =
 
 (** Setting states for building *)
 
-let get_universe tog st =
+let get_universe tog requested st =
   let OpamListCommand.{doc; test; dev_setup; dev; _} = tog in
   OpamSwitchState.universe st ~doc ~test ~dev_setup ~force_dev_deps:dev
-    ~requested:st.installed
+    ~requested
     Query
 
-let simulate_new_state tog st universe install names =
+let get_installed st atoms =
+  let open OpamPackage.Set.Op in
+  List.fold_left (fun (select, missing) atom ->
+      let installed =
+        OpamPackage.Set.filter (OpamFormula.check atom) st.installed
+      in
+      if OpamPackage.Set.is_empty installed then
+        (select, atom :: missing)
+      else (installed ++ select, missing))
+    (OpamPackage.Set.empty, []) atoms
+
+let dry_install tog st universe install =
   match OpamSolver.resolve universe
           (OpamSolver.request ~install ()) with
   | Success solution ->
     let new_st = OpamSolution.dry_run st solution in
-    print_solution st new_st names solution;
-    new_st, get_universe tog new_st
+    print_solution st new_st
+      (OpamPackage.Name.Set.of_list (List.map fst install))
+      solution;
+    let select, missing = get_installed new_st install in
```

Comment:
```suggestion
    let root = OpamFormula.packages_of_atoms new_st.install install in
```
with that, you don't need to move and externalise get_installed

Reply:
done in https://github.com/ocaml/opam/pull/5919/commits/25d2e2284fe707068df808245942b022705e934e